go/proxyd/daisychain.go

```diff
@@ -0,0 +1,485 @@
+				log.Error("error starting metrics server", "err", err)
+			}
+		}()
+	}
+
+	// To allow integration tests to cleanly come up, wait
+	// 10ms to give the below goroutines enough time to
+	// encounter an error creating their servers
+	errTimer := time.NewTimer(10 * time.Millisecond)
+
+	if config.Server.RPCPort != 0 {
+		go func() {
+			if err := srv.RPCListenAndServe(config.Server.RPCHost, config.Server.RPCPort); err != nil {
+				if errors.Is(err, http.ErrServerClosed) {
+					log.Info("RPC server shut down")
+					return
+				}
+				log.Crit("error starting RPC server", "err", err)
+			}
+		}()
+	}
+
+	<-errTimer.C
+	log.Info("started daisychain")
+
+	return func() {
+		log.Info("shutting down daisychain")
+		srv.Shutdown()
+		log.Info("goodbye")
+	}, nil
+}
+
+// checkChainIDs will check that each configured backend returns
+// the same chain id
+func (s *DaisyChainServer) checkChainIDs() (*big.Int, error) {
+	// send a chain id request to each node to ensure they are on the same chain
+	req, _ := ParseRPCReq([]byte(`{"id":"1","jsonrpc":"2.0","method":"eth_chainId","params":[]}`))
+	chainIds := []*hexutil.Big{}
+	for _, backend := range s.Backends() {
+		res, _ := backend.Forward(context.Background(), req)
+		str, ok := res.Result.(string)
+		if !ok {
+			return nil, errors.New("cannot fetch chainid on start")
+		}
+		chainId := new(hexutil.Big)
+		err := chainId.UnmarshalText([]byte(str))
+		if err != nil {
+			return nil, err
+		}
+		chainIds = append(chainIds, chainId)
+	}
+
+	if len(chainIds) == 0 {
+		panic("cannot fetch remote chain id")
+	}
+	chainId := chainIds[0].ToInt()
+	for _, id := range chainIds {
+		if id.ToInt().Cmp(chainId) != 0 {
+			log.Crit("mismatched chain ids detected", "chain-id", chainId, "other", id)
+		}
+	}
+	log.Info("detected chain id", "value", chainId)
+	if s.l2ChainID != nil {
+		if s.l2ChainID.Cmp(chainId) != 0 {
+			return nil, fmt.Errorf("mismatched chainids: expected %d, got %d", s.l2ChainID, chainId)
+		}
+	}
+	return chainId, nil
+}
+
+// HandleRPC is the top level RPC handler for the daisychain.
+// All RPC requests are routed through this handler.
+func (s *DaisyChainServer) HandleRPC(w http.ResponseWriter, r *http.Request) {
+	ctx := populateContext(w, r, s.authenticatedPaths)
+	if ctx == nil {
+		return
+	}
+
+	doRequest := func(ctx context.Context, req *RPCReq) (*RPCRes, bool) {
+		argType, ok := argTypes[req.Method]
+		if !ok {
+			return NewRPCErrorRes(req.ID, ErrParseErr), false
+		}
+
+		values, err := parsePositionalArguments(req.Params, argType)
+		if err != nil {
+			return NewRPCErrorRes(req.ID, fmt.Errorf("%s: %w", ErrParseErr, err)), false
+		}
+
+		argument, ok := parseRequestOptions(values)
+		if !ok {
+			return NewRPCErrorRes(req.ID, ErrParseErr), false
+		}
+
+		req, err = trimRequestOptions(req, values)
+		if err != nil {
+			return NewRPCErrorRes(req.ID, ErrParseErr), false
+		}
+
+		// TODO(tynes): perhaps a better approach is to attempt to classify the request
```

Comment:
I think that this TODO could help with making unit tests where the function is unit tested for the correct types